### PR TITLE
Support latest windows image 1.2.8  with 3ds Max 2019

### PR DESCRIPTION
**Pull request by matthew-rollitt**
Updated templates to handle the new windows version 1.2.8 which includes 3dsmax 2019 and refactored some templates since we can use configuration variables now.

### CODE REVIEW

**Comment by ascobie** [comment]
Seems ok, but need Christian to look over it as well.

File: ncj/3dsmax/vray-dr/job.template.json [hunk 1]
```diff
@@ -199,7 +210,7 @@
...
+                            "coordinationCommandLine":"cmd.exe /c if (\"[parameters('vrayRenderer')]\" == \"VRayAdv\") if (\"[parameters('maxVersion')]\" == \"3ds Max 2018\") cmd.exe /c start \"\" \"C:\\Autodesk\\3dsMax2018\\vrayspawner2018.exe\" \"-port=[parameters('vrayPort')]\" if (\"[parameters('maxVersion')]\" == \"3ds Max 2019\")   cmd.exe /c start \"\" \"C:\\Autodesk\\3dsMax2019\\vrayspawner2019.exe\" \"-port=[parameters('vrayPort')]\" if (\"[parameters('vrayRenderer')]\" == \"VRayRT\") if (\"[parameters('maxVersion')]\" == \"3ds Max 2018\")  cmd.exe /c start C:\\server\\3dsMax2018Vray\\bin\\vray.exe -server -portNumber=[parameters('vrayPort')])\"  if (\"[parameters('maxVersion')]\" == \"3ds Max 2019\") cmd.exe /c start C:\\server\\3dsMax2019Vray\\bin\\vray.exe -server -portNumber=[parameters('vrayPort')])\"",
                             "commonResourceFiles":[
```
Comment (ascobie) [inline ncj/3dsmax/vray-dr/job.template.json]: coordinationCommandLine in the DR template is impossible to make sense of. Can we pull this out into another script to make it easier to read?

Comment (matthew-rollitt) [reply ncj/3dsmax/vray-dr/job.template.json]: I think it's possible to pull them out.

File: ncj/3dsmax/scripts/3dsmax.ps1 [hunk 2]
```diff
@@ -209,17 +210,34 @@ mkdir -Force images > $null
...
+    if($env:3DSMAX_2018_EXEC -MATCH "cmdio")
+    {
+        $max_exec = $env:3DSMAX_2018_EXEC
+    }
+}
+ElseIf ($maxVersion -eq "3ds Max 2019")
 {
-    $max_exec = $env:3DSMAX_2018_EXEC
+        $max_exec = $env:3DSMAX_2019_EXEC
+        if(![System.IO.File]::Exists($max_exec))
```
Comment (smith1511) [inline ncj/3dsmax/scripts/3dsmax.ps1]: Would be good to be consistent.

File: ncj/3dsmax/scripts/3dsmax.ps1 [hunk 3]
```diff
@@ -209,17 +210,34 @@ mkdir -Force images > $null
...
+        {
+            Write-Host "3ds Max 2019 doesn't exist on this rendering image, please use a newer version of the rendering image."
+            exit 1
+        }
 }
-
+Else 
+{
+    Write-Host "No version of 3ds Max was selected."
```
Comment (smith1511) [inline ncj/3dsmax/scripts/3dsmax.ps1]: Default to latest?  2019

File: ncj/3dsmax/scripts/PrepareVrayServices.bat [hunk 4]
```diff
@@ -0,0 +1,19 @@
+@echo off
+IF %4=="VRayAdv" (
```
Comment (smith1511) [inline ncj/3dsmax/scripts/PrepareVrayServices.bat]: Would be good to assign variables at the top for clarity.

set VRAY_RENDERER=%4

Comment (smith1511) [reply ncj/3dsmax/scripts/PrepareVrayServices.bat]: Or use powershell

File: ncj/3dsmax/standard/job.parameters.json [hunk 5]
```diff
@@ -17,6 +17,9 @@
    "pathFile":{
       "value":"MyScene.mxp"
    },
+   "maxVersion":{
+      "value":"3ds Max 2019"
```
Comment (smith1511) [inline ncj/3dsmax/standard/job.parameters.json]: Remove prefix?  Just "2019"?

File: ncj/3dsmax/scripts/3dsmax.ps1 [hunk 6]
```diff
@@ -209,17 +210,34 @@ mkdir -Force images > $null
 
 # Render
 $max_exec = "3dsmaxcmdio.exe"
-if ($env:3DSMAX_2018 -and (Test-Path "$env:3DSMAX_2018"))
+If ($maxVersion -eq "3ds Max 2018")
```
Comment (smith1511) [inline ncj/3dsmax/scripts/3dsmax.ps1]: Just version

File: ncj/3dsmax/standard/job.template.json [hunk 7]
```diff
@@ -18,6 +18,17 @@
...
             }
         },
+        "maxVersion": {
+            "type": "string",
+            "defaultValue": "3ds Max 2019",
+            "metadata": {
+                "description": "The renderer to use"
+            },
+            "allowedValues":[
+                "3ds Max 2018",
```
Comment (smith1511) [inline ncj/3dsmax/standard/job.template.json]: Just version

File: ncj/3dsmax/vray-dr/job.parameters.json [hunk 8]
```diff
@@ -11,6 +11,9 @@
     "inputFilegroupSas":{
         "value":""
     },
+    "maxVersion":{
+      "value":"3ds Max 2019"
```
Comment (smith1511) [inline ncj/3dsmax/vray-dr/job.parameters.json]: Just version

File: ncj/3dsmax/vray-dr/job.template.json [hunk 9]
```diff
@@ -18,6 +18,17 @@
                 "description": "The pool id"
             }
         },
+        "maxVersion": {
+            "type": "string",
+            "defaultValue": "3ds Max 2019",
```
Comment (smith1511) [inline ncj/3dsmax/vray-dr/job.template.json]: Just version

File: ncj/3dsmax/vray-dr/job.template.json [hunk 10]
```diff
@@ -199,14 +216,18 @@
                         },
                         "multiInstanceSettings":{
                             "numberOfInstances":"[parameters('nodeCount')]",
-                            "coordinationCommandLine":"cmd.exe /c if \"[parameters('vrayRenderer')]\" == \"VRayAdv\" (cmd.exe /c start \"\" \"C:\\Autodesk\\3dsMax2018\\vrayspawner2018.exe\" \"-port=[parameters('vrayPort')]\") else (cmd.exe /c start C:\\server\\3dsmaxVray\\bin\\vray.exe -server -portNumber=[parameters('vrayPort')])",
+                            "coordinationCommandLine":"cmd /c %AZ_BATCH_JOB_PREP_WORKING_DIR%/PrepareVrayServices.bat --MaxVersion [parameters('maxVersion')] --VrayRenderer [parameters('vrayRenderer')] --VrayPort [parameters('vrayPort')])",
```
Comment (smith1511) [inline ncj/3dsmax/vray-dr/job.template.json]: Don't use these named args as you don't check for them in the batch file.  Just be clear in the batch file that the args are: <MaxVersion> <VRayRenderer> <VRayPort>

**Review by smith1511 — CHANGES_REQUESTED** [review]

File: ncj/3dsmax/vray-dr/job.template.json [hunk 11]
```diff
@@ -199,14 +216,18 @@
                         },
                         "multiInstanceSettings":{
                             "numberOfInstances":"[parameters('nodeCount')]",
-                            "coordinationCommandLine":"cmd.exe /c if \"[parameters('vrayRenderer')]\" == \"VRayAdv\" (cmd.exe /c start \"\" \"C:\\Autodesk\\3dsMax2018\\vrayspawner2018.exe\" \"-port=[parameters('vrayPort')]\") else (cmd.exe /c start C:\\server\\3dsmaxVray\\bin\\vray.exe -server -portNumber=[parameters('vrayPort')])",
+                            "coordinationCommandLine":"cmd /c %AZ_BATCH_JOB_PREP_WORKING_DIR%/PrepareVrayServices.bat [parameters('maxVersion')] [parameters('vrayRenderer')] [parameters('vrayPort')])",
```
Comment (smith1511) [inline ncj/3dsmax/vray-dr/job.template.json:219]: These args aren't in the correct order.  The script takes <VRay> <Max> <Port>.  I assume this hasn't been tested, can you test the changes before you merge please?

**Review by smith1511 — CHANGES_REQUESTED** [review]

File: ncj/3dsmax/vray-dr/job.template.json (hunk 11 above)
Comment (matthew-rollitt) [reply ncj/3dsmax/vray-dr/job.template.json:219]: Updated the bat file 👍

**Review by smith1511 — APPROVED** [review]

**Pull request merged**